### PR TITLE
PrometheusExporter: New concurrency handling for scrape middleware + http server

### PR DESCRIPTION
**Pull request by CodeBlanch**
## Changes

Removed the semaphore and added logic that returns the same response to concurrent scrape requests and then caches the response for a configurable period of time (10 seconds by default).

## Public API Changes

```csharp
public class PrometheusExporterOptions
{
   public int ScrapeResponseCacheDurationMilliseconds { get; set; }
}
```

## TODOs

* [X] `CHANGELOG.md` updated for non-trivial changes
* [X] `README.md` update for new setting
* [ ] Changes in public API reviewed

### CODE REVIEW

File: src/OpenTelemetry.Exporter.Prometheus/Implementation/PrometheusCollectionManager.cs
Comment (reyang) [inline src/OpenTelemetry.Exporter.Prometheus/Implementation/PrometheusCollectionManager.cs:28]: minor: consider the following name to align with https://github.com/open-telemetry/opentelemetry-dotnet/blob/5e369a6af1e0d7fd0d03bdd5035e7fe325ae4d98/src/OpenTelemetry/BatchActivityExportProcessor.cs#L27.

```suggestion
        private readonly int scrapeResponseCacheDurationMilliseconds;
```

File: src/OpenTelemetry.Exporter.Prometheus/Implementation/PrometheusExporterHttpServer.cs
Comment (reyang) [inline src/OpenTelemetry.Exporter.Prometheus/Implementation/PrometheusExporterHttpServer.cs:152]: This looks much cleaner 👍.

**Review by reyang — APPROVED** [review]

File: src/OpenTelemetry.Exporter.Prometheus/Implementation/PrometheusCollectionManager.cs
Comment (cijothomas) [inline src/OpenTelemetry.Exporter.Prometheus/Implementation/PrometheusCollectionManager.cs:53]: nit: incorrect comment about "10 seconds"

File: src/OpenTelemetry.Exporter.Prometheus/Implementation/PrometheusCollectionManager.cs
Comment (cijothomas) [inline src/OpenTelemetry.Exporter.Prometheus/Implementation/PrometheusCollectionManager.cs:59]: nit: for troubleshooting purposes, i think it'd be good to add some logging to indicate whether we re-used previous view, or triggered new collection.

**Review by cijothomas — APPROVED** [review]

File: src/OpenTelemetry.Exporter.Prometheus/Implementation/PrometheusCollectionManager.cs
Comment (CodeBlanch) [reply src/OpenTelemetry.Exporter.Prometheus/Implementation/PrometheusCollectionManager.cs:28]: Updated

File: src/OpenTelemetry.Exporter.Prometheus/Implementation/PrometheusCollectionManager.cs
Comment (CodeBlanch) [reply src/OpenTelemetry.Exporter.Prometheus/Implementation/PrometheusCollectionManager.cs]: Updated

File: src/OpenTelemetry.Exporter.Prometheus/Implementation/PrometheusSerializerExt.cs
Comment (CodeBlanch) [inline src/OpenTelemetry.Exporter.Prometheus/Implementation/PrometheusSerializerExt.cs:71]: @reyang I think this was a bug. We were getting nullref exception when keys/labels were not specified.

Comment (reyang) [reply src/OpenTelemetry.Exporter.Prometheus/Implementation/PrometheusSerializerExt.cs:71]: Yeah, I noticed this.

Comment (reyang) [reply src/OpenTelemetry.Exporter.Prometheus/Implementation/PrometheusSerializerExt.cs:71]: Let me add a test case for it once this PR is merged.

File: src/OpenTelemetry.Exporter.Prometheus/Implementation/PrometheusCollectionManager.cs
Comment (CodeBlanch) [reply src/OpenTelemetry.Exporter.Prometheus/Implementation/PrometheusCollectionManager.cs:60]: OK I was working on a log message but I'll hold off and do this as a follow-up. I like the idea of using the header.

File: src/OpenTelemetry.Exporter.Prometheus/Implementation/PrometheusSerializerExt.cs
Comment (reyang) [reply src/OpenTelemetry.Exporter.Prometheus/Implementation/PrometheusSerializerExt.cs:71]: #2612

File: src/OpenTelemetry.Exporter.Prometheus/Implementation/PrometheusSerializerExt.cs
Comment (mic-max) [inline src/OpenTelemetry.Exporter.Prometheus/Implementation/PrometheusSerializerExt.cs:80]: @CodeBlanch Would there be benefit to moving this to above the for loop, starting `i = 1` and also adding `cursor = WriteLabel(buffer, cursor, keys[i], values[i]);` or would the compiler already do this loop unrolling?

Comment (mic-max) [reply src/OpenTelemetry.Exporter.Prometheus/Implementation/PrometheusSerializerExt.cs:80]: Oops I misread the code. another idea would be to always add the comma character after the `WriteLabel` call and then decrement the cursor after the loop so the next statement that modifies the buffer will overwrite it.

File: src/OpenTelemetry.Exporter.Prometheus/Implementation/PrometheusCollectionManager.cs
Comment (mic-max) [inline src/OpenTelemetry.Exporter.Prometheus/Implementation/PrometheusCollectionManager.cs:163]: `this.exporter.OnExport = this.OnCollect;`
Would this not have worked?

Comment (CodeBlanch) [reply src/OpenTelemetry.Exporter.Prometheus/Implementation/PrometheusCollectionManager.cs:163]: So what this is doing is preventing the allocation of a delegate. The only sure-fire way I know how to do that is using a field. If you do it locally, sometimes the compiler will generate a field and initialize it on the first invocation, other times it will just allocate a delegate for each call. Not totally sure how it decides. I can't remember if this particular case fell into the delegate path or if I was just being cautious. I did do some benchmarks over this when I was working on it so I _probably_ saw allocations and used the field, but I can't remember exactly.

File: src/OpenTelemetry.Exporter.Prometheus/Implementation/PrometheusSerializerExt.cs
Comment (CodeBlanch) [reply src/OpenTelemetry.Exporter.Prometheus/Implementation/PrometheusSerializerExt.cs:80]: Cool idea! Updated on https://github.com/open-telemetry/opentelemetry-dotnet/pull/2749

File: src/OpenTelemetry.Exporter.Prometheus/Implementation/PrometheusCollectionManager.cs
Comment (mic-max) [reply src/OpenTelemetry.Exporter.Prometheus/Implementation/PrometheusCollectionManager.cs:163]: Thanks for the explanation, I had a hunch it was for perf reasons

**Pull request merged**
